### DISCUSSION TITLE
How can I add a comment to a query?

### DISCUSSION THREAD

**Discussion by Lewiscowles1986** [sqlalchemy/sqlalchemy#11115] [Usage Questions] [answered]
Hello and thank you for your work on SQLAlchemy. It certainly seems the most popular and robust way to interact with SQL in python.

I am looking to add one or more comments to specific queries, and I can't seem to find a place to do this. The reason I think a comment is ideal is that it is inert!

Right now I'm forced to use a CTE with literal values, and then filter my query to check that the generated CTE, which always has exactly one row, exists, in order to force the SQL into output.

I'm concerned this might further impact the performance I'm seeking to monitor.

We are using the RDS slow-query log, so the hope is that by adding these comments to the SQL we use, we will be able to get attribution for where they are coming from.

```
def label_query(name: str, **kwargs: str) -> CTE:
    """Use CTE to label queries using literal column."""
    return select(
        [
            literal(name).label("query name"),
        ]
        + [literal(kwargs[kwarg]).label(kwarg) for kwarg in kwargs]
    ).cte("query_label")


def force_apply_cte(query: Query, cte: CTE) -> Query:  # type: ignore[type-arg]
    """Force apply CTE to a query."""
    dummy_check = select(
        [cast(Integer, literal(1).label("query_label_used"))]
    ).select_from(cte)
    return query.filter(dummy_check.exists())
```

I would not be surprised if this didn't work outside of MySQL, but that is the specific variant of RDS we are using, so I'm happy our primary automated test suite passes and behaviour is the same using this query.

Once deployed to an environment we then performance test and I can certainly imagine situations where this could make performance even worse.

The usage of this (in each site we want to use it in) for-now is as follows:

```
def get_record(identifier: str, session: Session, **kwargs: str) -> SomeSqlAlchemyModel:
    """Return a some sql alchemy model."""
    try:
        query_label = label_query("get_record", **kwargs)
        return cast(
            SomeSqlAlchemyModel,
            force_apply_cte(
                session.query(SomeSqlAlchemyModel).where(
                    SomeSqlAlchemyModel.column == identifier,
                ),
                query_label,
            ).one(),
        )
```

Several things have been noted, such as the surprising necessity to add this to the right place; which is a little bit fragile and annoying; getting errors such as "can't add CTE after limit or order is applied", which we can work around addressing the sequence this is applied to.

Again, my primary concern with this approach at the moment is that if MySQL does not materialize the CTE it may be calculated over and over, per-record, which might suck performance, or ruin query-plans

**Comment by zzzeek** [comment, marked as answer]
we have a recipe for a system of injecting comments into SQL statements here:  https://github.com/sqlalchemy/sqlalchemy/wiki/SessionModifiedSQL

this makes use of the [before_cursor_execute](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.ConnectionEvents.before_cursor_execute) event hook to add SQL comments right into the raw SQL string.

**Reply by Lewiscowles1986** [reply]
If I were to want to do this at the query-level; do you think simply using that context manager might be enough? I'm guessing as long as I execute and then come out of that block, it should clean-up after itself?

**Reply by zzzeek** [reply]
that's how the recipe is made to work yes.

there's other ways to do this too, as this recipe is a bit old.  you could use a custom execution option too.  much simpler actually

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]


e = create_engine("sqlite://", echo=True)

@event.listens_for(e, "before_cursor_execute", retval=True)
def _apply_comment(connection, cursor, statement, parameters,
                                        context, executemany):
    if context and "custom_sql_comment" in context.execution_options:
        statement = f"\n-- {context.execution_options["custom_sql_comment"]}\n{statement}"
    return statement, parameters


Base.metadata.create_all(e)

with Session(e) as sess:
    sess.execute(
        select(A).execution_options(custom_sql_comment="my comment")
    )
```

**Reply by Lewiscowles1986** [reply]
This is really helpful. Thank you again. I'll try both as currently the code-base I'm in, is locked at January's 1.4 release. (we have a ticket to upgrade it, but sadly go fast vroom vroom, is more important right-now).

**Reply by Lewiscowles1986** [reply]
Sorry to be a pain in the backside, but it looks like some mysql clients strip comments before they are applied. I Understand this is not technically a sqlalchemy, but most likely a driver issue, but I do see it missing from the docs. Unsure how to solve right now, but if I work it out I'll come back and comment.

**Reply by Lewiscowles1986** [reply]
Oh, this is lovely, but dangerous. The SQL was not being escaped. A New-line made it through and a value with a semi-colon. So that was treated as a query.

I Guess with great power comes more responsibility... Why is there always responsibilty?
![Jack Sparrow Nope](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExOW1tY2h5MnY5dnE4dGt1ZDl5bWJrMDJhZ3N1ajhtcW52ZmlnZ2c0ZSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/5xtDarDazM7NnYvWIyQ/giphy.gif)
